Firestore/Source/Util/FSTDispatchQueue.mm

```diff
@@ -173,18 +182,37 @@ - (void)verifyIsCurrentQueue {
   FSTAssert([self onTargetQueue],
             @"We are running on the wrong dispatch queue. Expected '%@' Actual: '%@'",
             [self targetQueueLabel], [self currentQueueLabel]);
+  FSTAssert(_operationInProgress,
+            @"verifyIsCurrentQueue called outside enterCheckedOperation on queue '%@'",
+            [self currentQueueLabel]);
+}
+
+- (void)enterCheckedOperation:(void (^)(void))block {
+  FSTAssert(!_operationInProgress,
+            @"enterCheckedOperation may not be called when an operation is in progress");
+  @try {
+    _operationInProgress = YES;
+    [self verifyIsCurrentQueue];
+    block();
+  } @finally {
+    _operationInProgress = NO;
+  }
 }
 
 - (void)dispatchAsync:(void (^)(void))block {
-  FSTAssert(![self onTargetQueue],
+  FSTAssert(!_operationInProgress || ![self onTargetQueue],
```

Comment:
Hrm...  Why isn't this just `!_operationInProgress` ?  It would be invalid for _operationInProgress to be true when [self onTargetQueue] is not, but this assert makes it appear valid.